mantle/kola/tests/misc/network.go

```diff
@@ -219,6 +228,67 @@ var (
           umount /var/mnt
 	`
 
+	setupVethPairsTemplate = `#!/usr/bin/env bash
+          set -ex
+
+          create_veth_pair() {
+            veth_host_side_end_name=$1
+            veth_netns_side_end_name=$2
+            host_side_mac_address=$3
+            netns_side_ip=$4
+            network_namespace=$5
+
+            # Create veth pair and assign a namespace to veth-netns
+            ip link add ${veth_host_side_end_name} type veth peer name ${veth_netns_side_end_name}
+            ip link set ${veth_netns_side_end_name} netns ${network_namespace}
+
+            # Assign an MAC address to the host-side veth end
+            ip link set dev ${veth_host_side_end_name} address ${host_side_mac_address}
+
+            # Assign an IP address to netns-side veth end and bring it up
+            ip netns exec ${network_namespace} ip address add ${netns_side_ip} dev ${veth_netns_side_end_name}
+            ip netns exec ${network_namespace} ip -6 addr flush ${veth_netns_side_end_name}
+            ip netns exec ${network_namespace} ip link set ${veth_netns_side_end_name} up
+          }
+
+          activate_veth_end() {
+            veth_end_name=$1
+            nmcli dev set ${veth_end_name} managed yes
+            ip link set ${veth_end_name} up
+          }
+
+          primary_mac=%s
+          secondary_mac=%s
+          primary_ip=%s
+          secondary_ip=%s
+
+          network_namespace=test-netns
+          # Create a network namespace
+          ip netns add ${network_namespace}
+
+          # create 2 veth pairs between host side and network_namespace
+          veth1_host_side_end_name=veth1-host
+          veth1_netns_side_end_name=veth1-netns
+          create_veth_pair ${veth1_host_side_end_name} ${veth1_netns_side_end_name} ${primary_mac} 192.168.0.59 ${network_namespace}
+          veth2_host_side_end_name=veth2-host
+          veth2_netns_side_end_name=veth2-netns
+          create_veth_pair ${veth2_host_side_end_name} ${veth2_netns_side_end_name} ${secondary_mac} 192.168.0.60 ${network_namespace}
+
+          # Run a dnsmasq service on the network_namespace, to set the host-side veth ends a ip via their MAC addresses
+          echo -e "dhcp-range=192.168.0.50,192.168.0.60,255.255.255.0,12h\ndhcp-host=${primary_mac},${primary_ip}\ndhcp-host=${secondary_mac},${secondary_ip}" > /etc/dnsmasq.d/dhcp
```

Comment:
The server range overlaps with the server IP, this way we would give away the DHCP server IP. While the server should be within the subnet, I don't think it should be in the range

Reply:
not sure what you mean here. I don't see that the dnsmasq server got an ip, can you elabrate?:
```
dnsmasq: started, version 2.79 cachesize 150
dnsmasq: compile time options: IPv6 GNU-getopt DBus no-i18n IDN2 DHCP DHCPv6 no-Lua TFTP no-conntrack ipset auth DNSSEC loop-detect inotify
dnsmasq-dhcp: DHCP, IP range 192.168.0.50 -- 192.168.0.60, lease time 12h+ ip link set veth1-host up
dnsmasq: reading /etc/resolv.conf
dnsmasq: using nameserver 10.0.2.3#53
dnsmasq: read /etc/hosts - 2 addresses
dnsmasq-dhcp: DHCPDISCOVER(veth1-netns) 52:55:00:d1:56:00 
dnsmasq-dhcp: DHCPOFFER(veth1-netns) 192.168.0.55 52:55:00:d1:56:00 
dnsmasq-dhcp: DHCPREQUEST(veth1-netns) 192.168.0.55 52:55:00:d1:56:00 
dnsmasq-dhcp: DHCPACK(veth1-netns) 192.168.0.55 52:55:00:d1:56:00 
dnsmasq: reading /etc/resolv.conf
dnsmasq: using nameserver 10.0.2.3#53
dnsmasq: ignoring nameserver 192.168.0.59 - local interface
dnsmasq-dhcp: DHCPDISCOVER(veth2-netns) 52:55:00:d1:56:01 
dnsmasq-dhcp: DHCPOFFER(veth2-netns) 192.168.0.56 52:55:00:d1:56:01 
dnsmasq-dhcp: DHCPREQUEST(veth2-netns) 192.168.0.56 52:55:00:d1:56:01 
dnsmasq-dhcp: DHCPACK(veth2-netns) 192.168.0.56 52:55:00:d1:56:01 
dnsmasq: reading /etc/resolv.conf
dnsmasq: using nameserver 10.0.2.3#53
dnsmasq: ignoring nameserver 192.168.0.59 - local interface
dnsmasq: ignoring nameserver 192.168.0.60 - local interface
```

Reply:
Correct me if I'm wrong:

dhcp-range defines the range of IPs that will be served to clients
Your DHCP server is serving on interfaces with IPs 192.168.0.59 and 192.168.0.60 which happens to be in the range of offered IPs for clients. If you call the server with enough clients, you would assign 192.168.0.59 and 192.168.0.60 to some of the clients and that would lead to duplicated IPs on the network.

Reply:
oh I understand, you mean the netns-side veth ends. 
I figured we won't have any other dhcp requests, but you're right, better put them outside of the range.
Done